README.md
# connect-cas2

CAS Client NodeJS implement，support CAS 2.0+ protocol.

This project is forked from [connect-cas3](https://www.npmjs.com/package/connect-cas3) and have fixed some bugs.


## Install

    npm install node-cas2
            
## Quick start

```javascript
var express = require('express');
var ConnectCas = require('node-cas2');
var bodyParser = require('body-parser');
var session = require('express-session');
var cookieParser = require('cookie-parser');
var MemoryStore = require('session-memory-store')(session);

var app = express();

app.use(cookieParser());
app.use(session({
  name: 'NSESSIONID',
  secret: 'Hello I am a long long long secret',
  store: new MemoryStore()  // or other session store
}));

var casClient = new ConnectCas({
  debug: true,
    ignore: [
      /\/ignore/
    ],
    match: [],
    servicePrefix: 'http://10.17.86.87:8080',
    serverPath: 'http://rdmdev.oa.com',
    paths: {
      validate: '/cas/validate',
      serviceValidate: '/buglycas/serviceValidate',
      proxy: '/buglycas/proxy',
      login: '/buglycas/login',
      logout: '/buglycas/logout',
      proxyCallback: '/buglycas/proxyCallback'
    },
    redirect: false,
    gateway: false,
    renew: false,
    ssoff: true,
    cache: {
      enable: true,
      ttl: 5 * 60 * 1000,
      filter: []
    },
    fromAjax: {
      header: 'x-client-ajax',
      status: 418
    }
});

app.use(casClient.core());
app.get('/logout', casClient.logout());

// or do some logic yourself
app.get('/logout', function(req, res, next) {
  // Do whatever you like here, then call the logout middleware
  casClient.logout()(req, res, next);
});
```

## Constructor

```javascript

var casClient = new CasClient(options);

```

### options

#### options.serverPath {String} (Required)

The path of your CAS server. For example: https://www.your-cas-server-path.com/cas

#### options.servicePrefix {String} (Required)

The root path of your CAS client(Your website).

Every paths that for your CAS client will use this path as the root path.

For example: We will send `${options.servicePrefix}${options.paths.validate}` as `service` parameter to the login page. Then after login CAS server will redirect to `${options.servicePrefix}${options.paths.validate}` to validate the ST.

#### options.ignore {Array} (Optional, default: [])

In some cases, you don't need all request to be authenticated by the CAS. So you can set the ignore rules, when some rules matched, we will simply call the `next()` function and do nothing.

We support String rules, RegExp rules and Function rules.

For example, we checked the rules like:

1. String rules:

```javascript
if (req.path.indexOf(stringRule) > -1) next();
```

2. Reg rules:

```javascript
if (regRule.test(req.path)) next();
```

3. Function rules:
```javascript
if (funcRule(req.path, req)) next();
```

#### options.match {Array} (Optional, default: [])
If you set this option, only the paths that matched one of the rules will go into CAS middleware. All rules works just like `options.ignore`.

#### options.paths {Object} (Optional, default: {})
Relative paths to specific all functional paths to the CAS protocol. If you havn't modified the APIs of your CAS server, you may don't need to set this option. (In proxy mode, you must set the options.paths.proxyCall)

#### options.paths.validate (String) (Optional, default: '/cas/validate')
(For CAS Client)

The path you want your CAS client to validate ST from the CAS server. And we'll use `${options.servicePrefix}${options.paths.validate}` as `service` parameter to any CAS server's APIs that need this `service`.

#### options.paths.serviceValidate (String) (Optional, default: '/cas/serviceValidate')
(For CAS Server)

The path your CAS Server validate a ST. CAS client will send request to `${options.servicePrefix}${options.paths.serviceValidate}` to validate a ST.

#### options.paths.proxy (String) (Optional, default: '/cas/proxy')
(For CAS Server)

In proxy mode, you need a PGT(proxy granting ticket) to communicate with other server, you can get a PGT form this path: `${options.servicePrefix}${options.paths.proxy}`.

#### options.paths.login (String) (Optional, default: '/cas/login')
(For CAS Server)

The login path of your CAS server.

#### options.paths.logout (String) (Optional, default: '/cas/logout')
(For CAS Server)

The logout path of your CAS server.

#### options.paths.proxyCallback (String) (Optional, default: '')
(For CAS Client)
In proxy mode, setting this path means you want this path of your CAS Client to receive the callback from CAS Server(Proxy mode) to receive the PGTIOU and PGTID.

In none-proxy mode, don't set this option!

#### options.fromAjax {Object} (Optional, default: {})
Because an AJAX request can't notice 302 redirect response, and will directly redirect without telling you anything.

So when your user's authentication is expired, and your CAS Client and CAS Server is not in the same domain, (In most cases they won't be the same domain) when they send an AJAX request, oops, an not-allowed-cross-domain exception will occur!

To prevent this embarrassing situation, we add this option.

#### options.fromAjax.header {String} (Optional, default: 'x-client-ajax')
CAS will assume all request with header: 'x-client-ajax' is an AJAX request, so when user's authentication expired, CAS won't redirect to the login page, but send back the specified status code that you set as `options.fromAjax.status`.

#### options.fromAjax.status {Number} (Optional, default: 418)
As introduced before, when user's authentication expired, CAS won't redirect to the login page, but send back this as http status code.

For example:

    options.fromAjax = {
       header: 'x-client-ajax',
       status: 418
    };

So what you need to do in your browser code is you should handle this situation, when status code is 418, you should do `window.location.reload()` or something else to let your user reauthenticate.

#### options.debug {Boolean} (Optional, default: false)
For debug usage, we will log every step when CAS client is interacting with CAS server.

#### options.redirect(req, res) {Function} (Optional, default: null)
Change the default behaviour that after user login or login failed, CAS redirect to the last url.

If you return `true` from this redirect function, then CAS won't redirect to the last url, (So make sure you send a response in your function.)
otherwise, CAS just keep the same logic that will redirect to the last url after user login or login failed.

#### options.cache {Object} (Optional) `Since v1.1.0-beta`
Works in PROXY-MODE ONLY!

If you don't want to request a proxy-ticket every time you interact with your server, you can set this options to `enable=true`.
(NOTICE!! You must make sure the server you want to interact already set the PT cacheable, or using an PT that cached in your CAS client won't work!)

#### options.cache.enable {Boolean} (Optional, defualt: false)

#### options.cache.ttl {Number} (Optional, default: 300000 , in millisecond )
How long will the PT be cached.

#### options.cache.filter {Array} (Optinal, default: [])
In some cases, not every servers you want to interact enabled cacheable PT, so you can set the filter rules here, if one rule matchs the service url, it's PT won't be cached.

Every rule works like the `options.ignore`.

#### options.defaultLastUrl {String} (Optional, defualt: '/')

The default last url to redirect to when no record in session.

### METHOD

#### casClient.core()

Return a middleware that handles all the CAS client logic.

Use as `app.use(casClient.core())`.

#### casClient.logout()
Return a middleware that handles the logout action.

Use it like `app.get('/logout', casClient.logout())`. It will destroy the session of current user and then redirect to the CAS server logout page.

#### req.getProxyTicket(servicePath, [disableCache], callback)
If you setup options.paths.proxyCallback, CAS middleware will add a function called `getProxyTicket` on the request object, so you can access that by callback `req.getProxyTicket`.

`servicePath` is the service path you want to interact with.

`disableCache` {Boolean} (optional) When you setup PT cache, then you get a cached PT from `getProxyTicket`, but somehow the service validate this cached PT failed. In that case, you need to call this function again with setting this param to `true`.

`callback(err, pt)`

Example:
```javascript
   app.get('/api', function(req, res) {
     var service = 'http://your-service.com';
     req.getProxyTicket(service + '/cas', function(err, pt) {
       if (err) return res.status(401).send('Error when requesting PT, Authentication failed!');

       request.get(service + '/api/someapi?ticket=' + 'pt', function(err, response) {
          res.send(response);
       });
     });
   });
```

### PROXY MODE
In proxy mode, before you want to interact with your server, you need to request a ticket from CAS server.

### NONE-PROXY MODE
In none-proxy mode, don't set options.paths.proxyCallback, when all middle-ware passed, that means the login succeed.


## More

Currently we use this project in our production environment, and it works fine with our CAS Server. If you're facing any issues, please make us know!

## License

  MIT


CHANGELOG.md
## [0.0.3](https://github.com/knickknackfly/node-cas/compare/v0.0.2...v0.0.3) (2020-05-22)

* change `req.session.user` to `req.session.username`	([f98b1a4](https://github.com/knickknackfly/node-cas/commit/f98b1a4800926cf28640bed3add85da3fce1cc50))


## 0.0.2 (2020-05-21)


### Features

* add options.defaultLastUrl to config default redirect target ([9faa94d](https://github.com/knickknackfly/node-cas/commit/9faa94d3a0a04f424900785a6edf7da1cb1fc7ac))


## 0.0.1 (2020-05-07)

### Features

* clone from https://www.npmjs.com/package/connect-cas3
* store user info in session (req.session.user) ([a00ae5a](https://github.com/knickknackfly/node-cas/commit/a00ae5ad8e14f08cafac3ffe93ba8df61099584f))


lib/utils.js
var qs = require('query-string');
var url = require('url');
var http = require('http');
var https = require('https');
var buffer = require('buffer');

/**
 * Return `true` when pathname match the rule.
 *
 * @param req
 * @param pathname
 * @param rule
 * @returns {*}
 */
function isMatchRule(req, pathname, rule) {
  if (typeof rule === 'string') {
    return pathname.indexOf(rule) > -1;
  } else if (rule instanceof RegExp) {
    return rule.test(pathname);
  } else if (typeof rule === 'function') {
    return rule(pathname, req);
  }
}

/**
 *
 * @param req
 * @param options
 * @returns {string}
 */
function getOrigin(req, options) {
  var query = req.query;
  if (query.ticket) delete query.ticket;
  var querystring = qs.stringify(query);
  if (!options) {
    throw new Error('no options!!!');
  }

  return options.servicePrefix + url.parse(req.originalUrl).pathname + (querystring ? '?' + querystring : '');
}

/**
 * Check options.match first, if match, return `false`, then check the options.ignore, if match, return `true`.
 *
 * If returned `true`, then this request will bypass CAS directly.
 *
 * @param req
 * @param options
 */
function shouldIgnore(req, options) {
  if (options.match && options.match.splice) {
    var hasMatch = options.match.some(function(rule) {
      return isMatchRule(req, req.path, rule);
    });

    if (hasMatch) return false;
  }

  if (options.ignore && options.ignore.splice) {
    var hasMatchIgnore = options.ignore.some(function(rule) {
      return isMatchRule(req, req.path, rule);
    });

    if (hasMatchIgnore) return true;
  }
}

function getDefaultLastUrl(options) {
  return options.defaultLastUrl || '/'
}

function getLastUrl(req, options) {
  var lastUrl = (req.session && req.session.lastUrl) ? req.session.lastUrl : getDefaultLastUrl(options);

  var uri = url.parse(lastUrl);

  if (uri.pathname === options.paths.validate) lastUrl = getDefaultLastUrl(options);

  if (options.debug) {
    console.log('Get lastUrl: ' + lastUrl);
  }
  return lastUrl;
}

/**
 * Send a GET request
 *
 * @param path
 * @param callback
 */
function getRequest(path, callback) {
  sendRequest(path, {
    method: 'get'
  }, callback);
}

function postRequest(path, data, callback) {
  // callback(null, '<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN"><html><head><title>201 The request has been fulfilled and resulted in a new resource being created</title></head><body><h1>TGT Created</h1><form action="http://rdmdev.oa.com/buglycas/v1/tickets/TGT-3-o4VtFuR2Q7pddCckaNEaioRW0BoNTGCb4pV7mJI1PwdAvxCRGd-login.rdm.org?hi" method="POST">Service:<input type="text" name="service" value=""><br><input type="submit" value="Submit"></form></body></html>');
  sendRequest(path, {
    method: 'post',
    data: data
  }, callback);
}

function deleteRequest(path, callback) {
  sendRequest(path, {
    method: 'delete'
  }, callback);
}

function sendRequest(path, options, callback) {
  var requestOptions = url.parse(path);
  requestOptions.method = options.method;

  var isPost = options.method.toLowerCase() === 'post';

  if (isPost) {
    requestOptions.headers = {
      'Content-Type': 'application/x-www-form-urlencoded'
    };
    if (options.data) {
      var postData = [];
      for (var i in options.data) {
        postData.push(i + '=' + encodeURIComponent(options.data[i]));
      }
      options.data = postData.join('&');
      requestOptions.headers['Content-Length'] = Buffer.byteLength(options.data);
    }
  }

  var chunks = [];

  var req = (requestOptions.protocol === 'https:' ? https : http )['request'](requestOptions, function(res) {
    res.setEncoding('utf8');
    res.on('data', function(chunk) {
      chunks.push(chunk);
    });
    res.on('end', function() {
      callback(null, {
        status: res.statusCode,
        body: chunks.join(''),
        header: res.headers
      });
    });
  });

  req.on('error', function(e) {
    callback(e);
  });

  if (isPost && options.data) {
    req.write(options.data);
  }

  req.end();
}

function getPath(name, options) {
  var path = '';

  switch (name) {
    case 'login':
      path = options.serverPath + options.paths.login + '?service=' + encodeURIComponent(options.servicePrefix + options.paths.validate);
      break;
    case 'logout':
      path = options.serverPath + options.paths.logout + '?service=' + encodeURIComponent(options.servicePrefix + options.paths.validate);
      break;
    case 'pgtUrl':
      var proxyCallbackUri = url.parse(options.paths.proxyCallback);
      path = (proxyCallbackUri.protocol && proxyCallbackUri.host) ? (options.paths.proxyCallback) : (options.servicePrefix + options.paths.proxyCallback);
      break;
    case 'serviceValidate':
      path = options.serverPath + options.paths.serviceValidate;
      break;
    case 'proxy':
      path = options.serverPath + options.paths.proxy;
      break;
    case 'service':
    case 'validate':
      path = options.servicePrefix + options.paths.validate;
      break;
    case 'restletIntegration':
      path = options.serverPath + options.paths.restletIntegration;
  }
  return path;
}

function toArray(arrayLike) {
  return Array.prototype.slice.call(arrayLike);
}

module.exports = {
  toArray: toArray,
  getLastUrl: getLastUrl,
  getOrigin: getOrigin,
  shouldIgnore: shouldIgnore,
  deleteRequest: deleteRequest,
  getRequest: getRequest,
  postRequest: postRequest,
  getPath: getPath,
  isMatchRule: isMatchRule
};

lib/validate.js
var url = require('url');
var fetch = require('node-fetch');
var queryString = require('query-string');
var xml2js = require('xml2js').parseString;
var stripPrefix = require('xml2js/lib/processors').stripPrefix;
var http = require('http');
var utils = require('./utils');

/**
 * Validate a ticket from CAS server
 *
 * @param req
 * @param res
 * @param next
 * @param options
 * @param logger
 */
function validate(req, res, next, options, logger) {
  // check ticket first`
  var ticket = req.query && req.query.ticket || null;
  var session = req.session;
  logger.info('Start validating ticket...');

  if (ticket) {
    logger.info('Find ticket in query', ticket);
    if (session.st && session.st === ticket) {
      logger.info('Ticket in query is equal to the one in session, go next()');
      return next();
    }

    validateTicket(req, options, logger, function(err, response) {
      logger.info('Receive from CAS server, status: ' + response.status);
      if (response.status === 200) {
        validateCasResponse(req, res, response.body, options, next, logger);
      } else {
        logger.error('Receive response from cas when validating ticket, but request failed with status code: ' + response.status + '!');
        res.status(401).send({
          message: 'Receive response from cas when validating ticket, but request failed with status code: ' + response.status + '.'
        });
      }
    });
  } else {
    var lastUrl = utils.getLastUrl(req, options);
    logger.info('Can\' find ticket in query, redirect to last url: ' + lastUrl);
    res.redirect(302, lastUrl);
  }
}

module.exports = validate;

/**
 * Validate ticket from CAS server
 *
 * @param req
 * @param options
 * @param logger = {logError, logInfo}
 * @param callback
 */
function validateTicket(req, options, logger, callback) {
  var query = {
    service: utils.getPath('service', options),
    ticket: req.query.ticket
  };

  if (options.paths.proxyCallback) query.pgtUrl = utils.getPath('pgtUrl', options);

  var casServerValidPath = utils.getPath('serviceValidate', options) + '?' + queryString.stringify(query);

  logger.info('Sending request to: "' + casServerValidPath + '" to validate ticket.');

  utils.getRequest(casServerValidPath, function(err, response) {
    if (err) {
      logger.error('Error when sending request to CAS server, error: ', err.toString());
      logger.error(err);
      return callback(err);
    }

    callback(null, response);
  });
}

/**
 * Parse response XML from CAS server and do the rest of validation.
 *
 * @param req
 * @param res
 * @param casBody
 * @param options
 * @param next
 * @param logger = {logError, logInfo}
 */
function validateCasResponse(req, res, casBody, options, next, logger) {
  var ticket = req.query.ticket;

  xml2js(casBody, {
    explicitRoot: false,
    tagNameProcessors: [stripPrefix]
  }, function(err, serviceResponse) {
    if (err) {
      logger.error('Failed to parse CAS server response when trying to validate ticket.');
      logger.error(err);
      return res.status(500).send({
        message: 'Failed to parse CAS server response when trying to validate ticket.',
        error: err
      });
    }

    var success = serviceResponse && serviceResponse.authenticationSuccess && serviceResponse.authenticationSuccess[0];
    var user = success && success.user && success.user[0];
    var pgtIou = success && success.proxyGrantingTicket && success.proxyGrantingTicket[0];

    if (!serviceResponse) {
      logger.error('Invalid CAS server response.');
      return res.status(500).send({
        message: 'Invalid CAS server response, serviceResponse empty.'
      });
    }

    var lastUrl = utils.getLastUrl(req, options);

    if (!success) {
      logger.error('Receive response from CAS when validating ticket, but the validation is failed. Redirect to the last request url: ' + lastUrl);
      logger.error('Cas response:', serviceResponse);
      if (typeof options.redirect === 'function' && options.redirect(req, res, next)) {
        return;
      }

      return res.redirect(302, lastUrl);
    }

    req.session.st = ticket;
    req.session.username = user

    if (options.ssoff) {
      req.sessionStore.set(ticket, { sid: req.session.id, cookie: req.session.cookie }, function(err) {
        if (err) {
          logger.info('Trying to store ticket in sessionStore for ssoff failed!');
          logger.error(err);
        }
      });
    }

    req.session.cas = {};
    for (var casProperty in success) {
      if (casProperty != 'proxyGrantingTicket') {
        req.session.cas[casProperty] = success[casProperty][0];
      }
    }

    if (!pgtIou) {
      if (options.paths.proxyCallback) {
        logger.error('pgtUrl is specific, but havn\'t find pgtIou from CAS validation response! Response status 401.');
        return res.status(401).send({
          message: 'pgtUrl is specific, but havn\'t find pgtIou from CAS validation response!'
        });
      } else {
        logger.info('None-proxy mode, validate ticket succeed, redirecting to lastUrl: ' + lastUrl);
        req.session.save(function(err) {
          if (err) {
            logger.error('Trying to save session failed!');
            logger.error(err);
            return res.status(500).send({
              message: 'Trying to save session failed!',
              error: err
            });
          }
          if (typeof options.redirect === 'function' && options.redirect(req, res)) {
            logger.info('Specific options.redirect matched, redirect to customize location');
            return;
          }

          return res.redirect(302, lastUrl);
        });
      }

      return;
    }

    retrievePGTFromPGTIOU(req, res, pgtIou, options, logger);
  });
}

/**
 * Find PGT by PGTIOU
 *
 * @param req
 * @param res
 * @param pgtIou
 * @param options
 * @param logger
 */
function retrievePGTFromPGTIOU(req, res, pgtIou, options, logger) {
  logger.info('Trying to retrieve pgtId from pgtIou...');

  req.sessionStore.get(pgtIou, function(err, session) {
    if (err) {
      logger.error('Get pgtId from sessionStore failed!');
      logger.error(err);
      req.sessionStore.destroy(pgtIou);
      return res.status(500).send({
        message: 'Get pgtId from sessionStore failed!',
        error: err
      });
    }

    if (session && session.pgtId) {
      var lastUrl = utils.getLastUrl(req, options);

      req.session.pgt = session.pgtId;

      req.session.save(function(err) {
        if (err) {
          logger.error('Trying to save session failed!');
          logger.error(err);
          return res.status(500).send({
            message: 'Trying to save session failed!',
            error: err
          });
        }

        // 释放
        req.sessionStore.destroy(pgtIou);

        if (typeof options.redirect === 'function' && options.redirect(req, res)) {
          logger.info('CAS proxy mode login and validation succeed, pgtId finded. Customize redirect matched, redirect to comstomize location.');
          return;
        }

        logger.info('CAS proxy mode login and validation succeed, pgtId finded. Redirecting to lastUrl: ' + lastUrl);
        return res.redirect(302, lastUrl);
      });
    } else {
      logger.error('CAS proxy mode login and validation succeed, but can\' find pgtId from pgtIou: `' + pgtIou + '`, maybe something wrong with sessionStroe!');
      res.status(401).send({
        message: 'CAS proxy mode login and validation succeed, but can\' find pgtId from pgtIou: `' + pgtIou + '`, maybe something wrong with sessionStroe!'
      });
    }
  });
}


index.js
var _ = require('lodash');

var validate = require('./lib/validate');
var proxyCallback = require('./lib/proxyCallback');
var authenticate = require('./lib/authenticate');
var slo = require('./lib/slo');
var getProxyTicket = require('./lib/getProxyTicket');
var getProxyTicketThroughRestletReq = require('./lib/getProxyTicketThroughRestletReq');
var PTStroe = require('./lib/ptStroe');
var utils = require('./lib/utils');
var clearRestletTGTs = require('./lib/clearRestletTGTs');
var url = require('url');
var globalStoreCache = require('./lib/globalStoreCache');

var DEFAULT_OPTIONS = {
  debug: false,
  ignore: [],
  match: [],
  servicePrefix: '',
  serverPath: '',
  paths: {
    validate: '/cas/validate',
    serviceValidate: '/cas/serviceValidate',
    proxy: '/cas/proxy',
    login: '/cas/login',
    logout: '/cas/logout',
    proxyCallback: '/cas/proxyCallback'
  },
  redirect: false,
  gateway: false,
  renew: false,
  slo: true,
  // Is proxy-ticket cacheable
  cache: {
    enable: false,
    ttl: 5 * 60 * 1000, // In millisecond
    filter: []
  },
  fromAjax: {
    header: 'x-client-ajax',
    status: 418
  },
  // demo:
  // [{
  //   trigger: function(req) {
  //     return false;
  //   },
  //   params: {
  //     username: '',
  //     password: ''
  //   }
  // }]
  restletIntegration: {}
};

function ConnectCas(options) {
  if (!(this instanceof ConnectCas)) return new ConnectCas(options);

  this.options = _.merge({}, DEFAULT_OPTIONS, options);

  if (!this.options.servicePrefix || !this.options.serverPath) throw new Error('Unexpected options.service or options.serverPath!');

  this.ptStore = new PTStroe(_.merge({}, this.options.cache, { debug: this.options.debug }));

  this.logger = this.options.logger || function(req, type) {
      return console[type].bind(console[type]);
    };

  var pgtURI = url.parse(this.options.paths.proxyCallback);

  this.proxyCallbackPathName = (pgtURI.protocol && pgtURI.host) ? pgtURI.pathname : this.options.paths.proxyCallback;
}

ConnectCas.prototype.LoggerFactory = function() {
};

ConnectCas.prototype.core = function() {
  var options = this.options;
  var that = this;

  return function(req, res, next) {
    if (!req.sessionStore) throw new Error('You must setup a session store before you can use CAS client!');
    if (!req.session) throw new Error('Unexpected req.session ' + req.session);

    var pathname = req.path;
    var method = req.method;

    var logger = {
      info: that.logger(req, 'log'),
      error: that.logger(req, 'error'),
      warn: that.logger(req, 'warn'),
      log: that.logger(req, 'log')
    };

    var matchedRestletIntegrateRule;

    if (options.restletIntegration) {
      if (options.paths.restletIntegration) {
        for (var i in options.restletIntegration) {
          if (options.restletIntegration[i] && typeof options.restletIntegration[i].trigger === 'function' && options.restletIntegration[i].trigger(req)) {
            matchedRestletIntegrateRule = i;
            req.clearRestlet = clearRestletTGTs.bind(null, options, logger);
            break;
          }
        }
      } else {
        logger.warn("options.restletIntegration is set, but options.paths.restletIntegration is undefined! Maybe you forget to set all your paths.")
      }
    }

    if (options.paths.proxyCallback) {
      req.getProxyTicket = function(targetService, disableCache, callback) {
        return matchedRestletIntegrateRule ? getProxyTicketThroughRestletReq.call(that, req, targetService, matchedRestletIntegrateRule, options.restletIntegration[matchedRestletIntegrateRule], logger, typeof disableCache === 'function' ? disableCache : callback) :
          getProxyTicket.call(that, req, targetService, disableCache, logger, callback);
      };
    }

    if (matchedRestletIntegrateRule) return next();

    if (utils.shouldIgnore(req, options)) return next();

    if (method === 'GET') {
      switch (pathname) {
        case options.paths.validate:
          return validate(req, res, next, options, logger);
        case that.proxyCallbackPathName:
          return proxyCallback(req, res, next, options, logger);
        default:
          return authenticate(req, res, next, options, logger);
      }
    } else if (method === 'POST' && pathname === options.paths.validate && options.slo) {
      return slo(req, res, next, options, logger);
    }

    next();
  };
};

ConnectCas.prototype.logout = function() {
  var options = this.options;

  return function(req, res, next) {
    if (!req.session) {
      return res.redirect('/');
    }
    // Forget our own login session

    if (req.session.destroy) {
      req.session.destroy();
    } else {
      // Cookie-based sessions have no destroy()
      req.session = null;
    }

    // Send the user to the official campus-wide logout URL
    return res.redirect(utils.getPath('logout', options));
  };
};

ConnectCas.prototype.getPath = function(name, req) {
  return utils.getPath(name, this.options, req);
};

ConnectCas.prototype.destroyRestletIntegrations = function() {
  var stores = globalStoreCache.getAll();

  for (var i in stores) {
    utils.deleteRequest(utils.getPath('restletIntegration') + '/' + stores[i], function(err, response) {
      if (err || !response || response.statusCode != 200 || response.status != 200) return console.error('Error when delete restletIntegartion-created TGT!');
      console.log('Delete restletIntegartion-created TGT succeed!', stores[i]);
    });
  }
};

module.exports = ConnectCas;
